pyomo/core/expr/differentiate/differentiate.py

```diff
@@ -0,0 +1,80 @@
+from .diff_with_sympy import differentiate as sympy_diff
+from .diff_with_pyomo import reverse_sd, reverse_ad
+from pyutilib.enum import Enum
+from pyomo.core.kernel.component_map import ComponentMap
+
+
+DiffModes = Enum('sympy', 'reverse_symbolic', 'reverse_numeric')
```

Comment:
I think I would call this just "`Modes`".  To me, `DiffModes` is an awkward abbreviation (I read it as difference) and is redundant (the namespace `differentiate.Modes` makes it clear what the modes refer to).